Code: Layer panel refactor

barklund opened the pull request:
## Context

This refactors the code in the layer panel.

Note that with this PR the following naming is attempted:

- Element - an element in the story
- Group - a group of elements in the story
- Layer - a row in the layer panel, which can represent either an element or a group.

So no more _layer=element_ stuff.

## To-do

- [x] Refactor layer list
- [x] Combine layer and group components and generalize where applicable
- [ ] Add tests where relevant

## User-facing changes

_None_

## Testing Instructions

1. Check that layer panel works as before.

## Checklist

<!-- Check these after PR creation -->

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/googleforcreators/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/googleforcreators/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #11576

Commit:
Code: Moved layer panel to proper folder structure

Commit:
Refactored layer list

Commit:
Added custom hook

Commit:
Merge branch 'main' into code/11576-layer-refactor

Commit:
Merge branch 'main' into code/11576-layer-refactor

Commit:
Merge branch 'main' into code/11576-layer-refactor

Commit:
Merge branch 'main' into code/11576-layer-refactor

Commit:
Renamed components in the layer list

Commit:
Completely refactored layers in the panel

miina commented on packages/story-editor/src/components/panels/layer/constants.js:
```diff
@@ -16,3 +16,5 @@
 export const LAYER_HEIGHT = 35;
 export const NESTED_PX = 30;
+
+export const nestedOffsetCalcFunc = (evt) => evt.offsetY < LAYER_HEIGHT / 1.7;
```
<!-- trivial preference-based requests -->
**🥜 nitpick:**
Should this maybe go to an utils file since it doesn't seem to be a constant? 😄

barklund replied:
Naa, that's too much trouble IMO. It's a constant function! 😂

miina commented on packages/story-editor/src/components/panels/layer/useLayers.js:
```diff
@@ -14,30 +14,107 @@
+    );
+  }, [elementLayers]);
+
+  const handlePositionChange = useCallback(
```
<!-- potential concern, ask for clarification or investigation -->
**❓question:**
I know it was brought over from another file and is unchanged, but does `handlePositionChange` here apply only when dragging a layer? Or could this be reapplied for keyboard events, too? Thinking that if it's just for dragging then would be good to specify the name for clarity (since we have #11906 for the keyboard handling).

barklund replied:
Renamed it - it is specifically for dragging only to handle the partial overlap and figure out what the new position is going to be.

miina reviewed (commented):
Generally looks great, nice refactor!

Looks like the count of layers is missing now (see the Percy screenshots for example, it shows the diff well, verified on local, too).

Commit:
Renamed callback

Commit:
Merge branch 'main' into code/11576-layer-refactor

Commit:
Fixed layer button notification

Layer button count should reflect the number of elements on the page, not the number of layers in the layer panel, as groups also count as layers, but are not actually elements.

barklund commented on packages/story-editor/src/components/footer/layers/layers.js:
```diff
@@ -42,7 +42,9 @@ const StyledNavigationWrapper = styled(NavigationWrapper)`
-  const layersLength = useLayers().length;
+  const nunElements = useStory(
+    ({ state }) => state.currentPage?.elements?.length
+  );
```
@miina, well-spotted - I didn't realize this used the same hook before. Well, now that layers mean all the rows in the layer panel (both elements and groups), I feel it makes sense for this to just show the number of elements rather than the full number of layers, or?

miina replied:
Makes sense to me!

miina commented on packages/story-editor/src/components/footer/layers/layers.js:
```diff
@@ -42,7 +42,9 @@ const StyledNavigationWrapper = styled(NavigationWrapper)`
 
 function Layers() {
-  const layersLength = useLayers().length;
+  const nunElements = useStory(
```
Nun elements? 😆 

Edit:

```suggestion
  const numElements = useStory(
```

barklund replied:
🙏 ⛪

barklund replied:
(fixed in 974f60f161)

miina reviewed (approved):
`nun` -> `num` and LGTM! ⛪

Commit:
Typo

Commit:
Merge branch 'main' into code/11576-layer-refactor

merapi committed:
Merge branch 'main' into code/11576-layer-refactor

Commit:
Merge branch 'main' into code/11576-layer-refactor

Commit:
Merge branch 'code/11576-layer-refactor' of github.com:GoogleForCreators/web-stories-wp into code/11576-layer-refactor

merapi committed:
Merge branch 'main' into code/11576-layer-refactor

merapi committed:
Merge branch 'main' into code/11576-layer-refactor
